src/threeIdProvider.ts

```diff
@@ -19,94 +20,98 @@ type Context = {
-      to: params.to,
-    })
+  '3id_encrypt': async ({ origin, keyring, permissions }, params) => {
+    if (!permissions.has(origin, params.spaces)) {
+      throw new RPCError(0, 'Authentication required')
+    }
+    const { to, blockSize, message, space } = params
+    const paddedMsg = typeof message === 'string' ? pad(message, blockSize) : message
+    if (to) {
+      return keyring.asymEncrypt(paddedMsg, to)
+    } else {
+      return keyring.symEncrypt(paddedMsg, { space })
+    }
   },
-  '3id_decrypt': async (ctx, params) => {
-    return await ctx.provider.wallet.decrypt(
-      {
-        ciphertext: params.ciphertext,
-        ephemeralFrom: params.ephemeralFrom,
-        nonce: params.nonce,
-      },
-      params.space,
-      params.buffer
-    )
+  '3id_decrypt': async ({ origin, keyring, permissions }, params) => {
+    if (!permissions.has(origin, params.spaces)) {
+      throw new RPCError(0, 'Authentication required')
+    }
+    const { ciphertext, ephemeralFrom, nonce, space } = params
+    let paddedMsg
+    if (ephemeralFrom) {
+      paddedMsg = keyring.asymDecrypt(
+        ciphertext,
+        ephemeralFrom,
+        nonce,
+        // @ts-ignore issue: https://github.com/microsoft/TypeScript/issues/14107
+        { space }
+      )
+    } else {
+      paddedMsg = keyring.symDecrypt(ciphertext, nonce, { space })
+    }
+    if (!paddedMsg) throw new RPCError('Could not decrypt message')
+    return unpad(paddedMsg)
   },
-  '3id_hashEntryKey': async (ctx, params) => {
-    return await ctx.provider.wallet.hashDBKey(params.key, params.space)
+  '3id_hashEntryKey': async ({ origin, keyring, permissions }, params) => {
+    if (!permissions.has(origin, params.spaces)) {
+      throw new RPCError(0, 'Authentication required')
+    }
+    const salt = keyring.getDBSalt(params.space)
+    return sha256Multihash(salt + params.key)
   },
-  '3id_newAuthMethodPoll': (ctx) => ctx.provider.pollAuthMethods(),
-  '3id_newLinkPoll': (ctx) => ctx.provider.pollLinks(),
+  '3id_newAuthMethodPoll': () => [],
+  '3id_newLinkPoll': () => [],
 }
 
 type Callback = (err: Error | null | undefined, res?: RPCResponse | null) => void
 
 export default class ThreeIdProvider {
-  private _newAuthMethods: Array<EncryptedMessage> = []
-  private _newLinks: Array<LinkProof> = []
-
   protected _handle: RequestHandler
 
-  public wallet: IdentityWallet
-
-  constructor(wallet: IdentityWallet) {
-    this._handle = createHandler<Context>(methods)
-    this.wallet = wallet
-
-    wallet.events.on('new-auth-method', (authBlob: EncryptedMessage) => {
-      this._newAuthMethods.push(authBlob)
-    })
-    wallet.events.on('new-link-proof', (linkProof: LinkProof) => {
-      this._newLinks.push(linkProof)
-    })
+  constructor({ permissions, threeIdx, keyring, forcedOrigin }: ProviderConfig) {
+    this._handle = (origin: string, req: RPCRequest) => {
+      return createHandler<Context>(Object.assign(methods, didMethods))({
```

Comment:
As commented above, it's not ideal to call `createHandler` for every call to handle